cli/internal/cmd/apply.go

```diff
@@ -0,0 +1,505 @@
+	                                  │                          ───┐
+	                    ┌─────────────▼────────────┐                │
+	     Can be skipped │Upgrade NodeVersion object│                │K8s/Image
+	  if we ran Init RP │  (Image and K8s update)  │                │Phase
+	                    └─────────────┬────────────┘                │
+	                                  │                          ───┘
+	                        ┌─────────▼──────────┐
+	                        │Write success output│
+	                        └────────────────────┘
+*/
+func (a *applyCmd) apply(cmd *cobra.Command, configFetcher attestationconfigapi.Fetcher, upgradeDir string) error {
+	// Read user's config and state file
+	a.log.Debugf("Reading config from %s", a.flags.pathPrefixer.PrefixPrintablePath(constants.ConfigFilename))
+	conf, err := config.New(a.fileHandler, constants.ConfigFilename, configFetcher, a.flags.force)
+	var configValidationErr *config.ValidationError
+	if errors.As(err, &configValidationErr) {
+		cmd.PrintErrln(configValidationErr.LongMessage())
+	}
+	if err != nil {
+		return err
+	}
+
+	a.log.Debugf("Reading state file from %s", a.flags.pathPrefixer.PrefixPrintablePath(constants.StateFilename))
+	stateFile, err := state.ReadFromFile(a.fileHandler, constants.StateFilename)
+	if err != nil {
+		return err
+	}
+
+	// Check license
+	a.log.Debugf("Running license check")
+	checker := license.NewChecker(a.quotaChecker, a.fileHandler)
+	if err := checker.CheckLicense(cmd.Context(), conf.GetProvider(), conf.Provider, cmd.Printf); err != nil {
+		cmd.PrintErrf("License check failed: %v", err)
+	}
+	a.log.Debugf("Checked license")
+
+	// Check if we already have a running Kubernetes cluster
+	// by checking if the Kubernetes admin config file exists
+	// If not, we need to run the init RPC first
+	a.log.Debugf("Checking if %s exists", a.flags.pathPrefixer.PrefixPrintablePath(constants.AdminConfFilename))
+	initRequired := false
+	if _, err := a.fileHandler.Stat(constants.AdminConfFilename); err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return fmt.Errorf("checking for %s: %w", a.flags.pathPrefixer.PrefixPrintablePath(constants.AdminConfFilename), err)
+		}
+		// Only run init RPC if we are not skipping the init phase
+		// This may break things further down the line
+		// It is the user's responsibility to make sure the cluster is in a valid state
+		initRequired = true && !a.flags.skipPhases.contains(skipInitPhase)
+	}
+	a.log.Debugf("Init RPC required: %t", initRequired)
+
+	// Validate input arguments
+
+	// Validate Kubernetes version as set in the user's config
+	// If we need to run the init RPC, the version has to be valid
+	// Otherwise, we are able to use an outdated version, meaning we skip the K8s upgrade
+	a.log.Debugf("Validating Kubernetes version %s", conf.KubernetesVersion)
+	validVersion, err := versions.NewValidK8sVersion(string(conf.KubernetesVersion), true)
```

Comment:
nit: I'd put the logic for handling init / upgrade for version validation in a function for readability

Reply:
if the K8s phase is skipped we might want to skip the validation